external PSK mode

kazuho opened the pull request:
Subsumes #321, https://github.com/sshock/picotls/pull/1.

ToDo:
* [ ] add end-to-end test against openssl?

kazuho commented on lib/picotls.c:
```diff
@@ -2567,6 +2603,10 @@ static int decode_server_hello(ptls_t *tls, struct st_ptls_server_hello_t *sh, c
+                              if (tls->ctx->key_exchanges == NULL) {
+                                  ret = PTLS_ALERT_HANDSHAKE_FAILURE;
+                                  goto Exit;
+                              }
```
Do we need to do something similar for negotiated groups?

kazuho commented on lib/picotls.c:
```diff
@@ -3998,11 +4042,15 @@ static int vec_is_string(ptls_iovec_t x, const char *y)
+ */
 static int try_psk_handshake(ptls_t *tls, size_t *psk_index, int *accept_early_data, struct st_ptls_client_hello_t *ch,
-                             ptls_iovec_t ch_trunc)
+                             ptls_iovec_t ch_trunc, const struct st_ptls_external_psk_t *external_psk)
```
We do not need an argument, as the values can be acceted via `tls->ctx`?

kazuho replied:
Fixed in e3e11fc.

kazuho commented on lib/picotls.c:
```diff
@@ -4372,145 +4434,156 @@ static int server_handle_hello(ptls_t *tls, ptls_message_emitter_t *emitter, ptl
+    }
+
+    /* try external psk handshake */
+    if (!is_second_flight && ch->psk.hash_end != 0 &&
```
Don't we need to do HRR if the client provides an empty KeyShare, negotiated groups, and an external PSK, with ke_modes set to PSK_DHE?

kazuho commented on t/picotls.c:
```diff
@@ -1834,6 +1947,7 @@ static void test_all_handshakes_core(void)
         subtest("stateless-hrr-aad-change", test_stateless_hrr_aad_change);
     }
     subtest("key-update", test_key_update);
+    subtest("pre-shared-key", test_pre_shared_key);
```
TODO: Find out how to test HRR.

kazuho reviewed (commented):
Review to self:

sshock commented:
As part of this work, could we add the ability to force PSK-only mode?  I found the `require_dhe_on_psk` option, but that only presents two scenarios, not three:

1. require_dhe_on_psk = 1:
    * client sends just psk_dhe_ke mode (and a key_share of course)
    * server chooses PSK-DHE of course and sends back a key_share
2. require_dhe_on_psk = 0:
    * client sends both psk_dhe_ke and psk_ke mode and a key_share
    * server chooses psk_ke (PSK-only) and does not send back a key_share
3. ??? <-- this isn't possible yet
    * client sends just psk_ke mode and no key_share
    * server chooses psk_ke (PSK-only) and does not send back a key_share

A couple reasons why this could be useful:
* Other TLS servers might choose psk_dhe_ke over psk_ke when presented with both modes.
* We can save time by not generating a key_share on the client that won't be used.

Thanks!

sshock commented:
@kazuho I noticed the following behavior with external PSK authentication:
* When the PSK identity matches but the secret is wrong, the server fails right away with error 51 PTLS_ALERT_DECRYPT_ERROR.  This is good.
* When the PSK identity doesn't match, the server continues the handshake, although the ServerHello lacks a selected pre_shared_key.  Then the client fails with error 47 PTLS_ALERT_ILLEGAL_PARAMETER.  Is this the expected behavior? 

I'm concerned that the failure is happening on the client.  If the client tried to keep going would it fail on the server?  I'm concerned because it appears from the server's perspective that the handshake completed since it did send a TLS Handshake Finished and there is a SERVER_TRAFFIC_SECRET_0.

kazuho commented:
@sshock
> I noticed the server is entering "force a full handshake" mode, which makes sense to do when no PSK is selected, and it is sending an empty EE, an empty Certificate, Finished, and then a New Session Ticket.
> 
> I still think it is not good that the server reaches handshake complete. I think if it reaches a point where it is wanting to send a 0-length Certificate, it probably ought to have failed by then.

Thank you for bringing the issue to our attention, I think it depends.

[RFC 8446 appendix C.5](https://datatracker.ietf.org/doc/html/rfc8446#appendix-C.5) states that _applications MUST NOT use TLS in such a way absent explicit configuration or a specific application profile._ To paraphrase, if a server can run anonymously depends on the intent of the user or the application, and the willingness of the TLS stack to support it.

Also, [Section 6.2](https://datatracker.ietf.org/doc/html/rfc8446#section-6.2) explicitly states that servers may avoid sending `unknown_psk_identity` alert due to privacy concerns.

As a solution, how about doing something like below inside `client_handle_hello`?
```c
if (tls->pre_shared_key.identity.base != NULL && !tls->is_psk_handshake) {
    ret =  PTLS_ALERT_HANDSHAKE_FAILURE;
    goto Exit;
}
```

Running as a PSK client, we know that the handshake cannot conclude when the server has rejected the PSK identity being offered.

sshock commented:
> @sshock
> 
> [RFC 8446 appendix C.5](https://datatracker.ietf.org/doc/html/rfc8446#appendix-C.5) states that _applications MUST NOT use TLS in such a way absent explicit configuration or a specific application profile._ To paraphrase, if a server can run anonymously depends on the intent of the user or the application, and the willingness of the TLS stack to support it.

Interesting.  I had not expected TLS to give any allowance for establishing an unauthenticated connection.

Although the scenarios described here seem to be related to situations where perhaps this might be the desired behavior, such as when using raw public keys.  I don't think the use case of "I'm a server using PSK for authentication; but if the client doesn't have a matching PSK, I want to let them connect anyways with no authentication" is a desired behavior, especially not by default.

> Also, [Section 6.2](https://datatracker.ietf.org/doc/html/rfc8446#section-6.2) explicitly states that servers may avoid sending `unknown_psk_identity` alert due to privacy concerns.

Oh neat.  I never even noticed the `unknown_psk_identity` error code before.

So this is what I see in section 6.2:
>    unknown_psk_identity:  Sent by servers when PSK key establishment is
>       desired but no acceptable PSK identity is provided by the client.
>       Sending this alert is OPTIONAL; servers MAY instead choose to send
>       a "decrypt_error" alert to merely indicate an invalid PSK
>       identity.
> 

I don't think that is saying a server can choose to send "unknown_psk_identity" or just not fail and keep going.  I think it is saying you can choose to send "decrypt_error" instead of "unknown_psk_identity".

> As a solution, how about doing something like below inside `client_handle_hello`?
> 
> ```c
> if (tls->pre_shared_key.identity.base != NULL && !tls->is_psk_handshake) {
>     ret =  PTLS_ALERT_HANDSHAKE_FAILURE;
>     goto Exit;
> }
> ```

Unfortunately, this won't work because the on_client_hello callback gets called before the server does psk handshake (so is_psk_handshake will just always be false in here).

> Running as a PSK client, we know that the handshake cannot conclude when the server has rejected the PSK identity being offered.

Correct.  The client side can and does fail when the server does not select a PSK.  Also, I am confident the connection would fail with a malicious server trying to continue by selecting a PSK even though it doesn't know it, as the server would not be able to derive the same handshake or master secret as the client.

I believe we only have an issue on the server side here.  We have set up a server to use PSK authentication and yet it will allow any client to connect without knowing the PSK secret by simply providing an incorrect PSK identity.  Instead of completing the handshake I think the server needs to fail with unknown_psk_identity or decrypt_error.

kazuho commented:
Thanks, I'm convinced. What do you think about #489?

sshock commented:
> Thanks, I'm convinced. What do you think about #489?

Cool, thanks!

Just one thing: did you mean to include the check for `require_client_authentication`?  I don't mind having to specify that to get the desired behavior on the server, but if we do want to use that a few spots will need to be tweaked for it to not cause problems:
* Line 4490; currently it completely skips external psk handshake if `require_client_authentication` is set.
* Similar issue on line 4595?
* Line 4615 will force us always into HANDSHAKE_MODE_FULL, which we don't want.
* Line 4760 will send a PTLS_HANDSHAKE_TYPE_CERTIFICATE_REQUEST
* `server_finish_handshake()` will go to state PTLS_STATE_SERVER_EXPECT_CERTIFICATE and return PTLS_ERROR_IN_PROGRESS

Maybe it would be better to have a new flag for this instead; then we could also invert the name and logic so the default behavior will be to require client authentication; e.g., call it `psk_allow_unauthenticated_client`.

kazuho commented:
@sshock That's a good point. I did not take that path because I wasn't certain to have one flag change the behavior of both external PSK mode and X.509 mode.

For external PSK, the only reason for having a flag is to allow fallback to authentication using other methods. For external PSK to allow fallbacks, the unified flag will be set to false. That means that users cannot use certificate-based client authentication as a fallback.

Is that the desired behavior?

Maybe it is not. Hence I ended up with a PR that does not change how `require_client_authentication` works at all.

Regarding these fall backs, rather than adding a flag to allow mismatch of keys, maybe we should better extend the `on_client_hello_cb` callback to carry the PSK identity so that application can switch to a different `ptls_context_t` that does not have the external PSK set. Adding flags lead to complexity.

kazuho commented:
> Regarding these fall backs, rather than adding a flag to allow mismatch of keys, maybe we should better extend the `on_client_hello_cb` callback to carry the PSK identity so that application can switch to a different `ptls_context_t` that does not have the external PSK set. Adding flags lead to complexity.

Did this and merged #489 here.

sshock commented:
> > Regarding these fall backs, rather than adding a flag to allow mismatch of keys, maybe we should better extend the `on_client_hello_cb` callback to carry the PSK identity so that application can switch to a different `ptls_context_t` that does not have the external PSK set. Adding flags lead to complexity.

Yeah, I agree that adding flags adds complexity.  And enhancing the on_client_hello callback is more flexible, although it is a breaking API change, so that's unfortunate.

So it's ok to for the on_client_hello callback to change the ptls_context_t?  That feels a little strange for one to be changing that in the middle of a handshake, but I guess it makes sense for this use case.

So an example use case here might be that in this callback one manually checks to see if there is at least one identity and it matches one of the psk identities on the server, and if not change the context to one that doesn't have external PSK to fallback to cert-based auth?

> Did this and merged #489 here.

Thanks.  Unfortunately, I see you still are still checking require_client_authentication on line 4601

```
    /* If the server was setup to use an external PSK but failed to agree, abort the handshake. Because external PSK is a form of
     * mutual authentication, we should abort continue the handshake upon negotiation failure, at least by default. */
    if (tls->ctx->pre_shared_key.identity.base != NULL && tls->ctx->require_client_authentication && psk_index == SIZE_MAX) {
        ret = PTLS_ALERT_UNKNOWN_PSK_IDENTITY;
        goto Exit;
    }
```

You can't check `require_client_authentication` here unless you also fix all the other spots I mentioned.  You're making it so we have to set `require_client_authentication` to get the desired behavior here, but setting that totally breaks PSK in multiple ways because all those other spots will do the wrong thing.

Do you understand how each of those spots needs to be fixed?  The first two spots I mentioned probably just need to take out the check for `!tls->ctx->require_client_authentication`.  All the other spots that check it I think need to be changed from just `tls->ctx->require_client_authentication` to `(tls->ctx->require_client_authentication && tls->ctx->pre_shared_key.identity.base == NULL)`.

kazuho commented:
@sshock Thanks, I missed that you were talking about me accidentally referring to `require_client_authentication`. I've addressed that in e220cda.

> So it's ok to for the on_client_hello callback to change the ptls_context_t? That feels a little strange for one to be changing that in the middle of a handshake, but I guess it makes sense for this use case.

This is actually pretty common practice. This is how we switch between different certificates based on SNI, not only when using picotls but also with other TLS stacks that I know of.

sshock commented:
Hi @kazuho.  I'm seeing an unexpected behavior for no matching PSK identity in a scenario where there is no key share.

In my testing earlier with mismatching PSK identity, the server would fail with a 115 UNKNOWN_PSK_IDENTITY as expected, as we fixed it here to do.

However, in my latest testing I have no key share (ptls_context_t key_exchanges is set to NULL), and now instead of failing with 115 as expected, the server sends a `HelloRetryRequest`.

This is happening around line 4633 in picotls.c, right after the `/* try external psk handshake */` section, where it does the HelloRetryRequest:
```
    /* send HelloRetryRequest if enforced by config or upon key-share mismatch, unless PSK has already been selected */
    if (!is_second_flight && psk_index == SIZE_MAX &&
        (key_share.algorithm == NULL || (properties != NULL && properties->server.enforce_retry))) {
```

I'm not sure what the right solution is, but it seems like maybe this block of code to fail with UNKNOWN_PSK_IDENTITY needs to be moved up (to right after the `/* try external psk handshake */` section)?

```
    /* If the server was setup to use an external PSK but failed to agree, abort the handshake. Because external PSK is a form of
     * mutual authentication, we should abort continue the handshake upon negotiation failure, at least by default. */
    if (tls->ctx->pre_shared_key.identity.base != NULL && psk_index == SIZE_MAX) {
        ret = PTLS_ALERT_UNKNOWN_PSK_IDENTITY;
        goto Exit;
    }
```

I'm not confident though because that is currently placed after a `/* try psk handshake */` section (is this second one for handling resumption psk?), so if we moved it up would it break resumption psk?

I'm hoping you can understand the proper way to fix this 🤞, thanks!

sshock commented:
Since my primary concern is security, I can live with this inconsistent behavior as long as it doesn't present a security vulnerability.  That is, as long as the 2nd ClientHello attempt will fail, both with or without a key share.  I just don't want a client to be able to trick a server into allowing an anonymous connection.

If you can help me confirm this, that would make me feel a lot better.

Then we can take our time thinking about the proper way to fix it, which I need your help with.

Or if you believe this behavior is correct, please explain.  The RFC says Hello Retry is when "it is able to find an acceptable set of parameters but the ClientHello does not contain sufficient information to proceed with the handshake."  I just don't see how "no matching PSK identity" falls into this category, but only when key share is missing.

kazuho commented:
@sshock
> I'm seeing an unexpected behavior for no matching PSK identity in a scenario where there is no key share.
> 
> In my testing earlier with mismatching PSK identity, the server would fail with a 115 UNKNOWN_PSK_IDENTITY as expected, as we fixed it here to do.
> 
> However, in my latest testing I have no key share (ptls_context_t key_exchanges is set to NULL), and now instead of failing with 115 as expected, the server sends a HelloRetryRequest.

Thank you for the patience.

I've made some changes reducing differences from master so that this PR can be merged. I think I've fixed the bug alongside that. We also have tests that involve PSK + HRR.

sshock commented:
> I've made some changes reducing differences from master so that this PR can be merged. I think I've fixed the bug alongside that. We also have tests that involve PSK + HRR.

Thanks @kazuho !  I was able to confirm that it now fails properly with 115 on wrong psk identity, even with no key share.

sshock commented:
I also noticed that cli.c now helps set the psk_hash (a522cd9), thanks!
